### PR TITLE
More namespace alignment

### LINKED ISSUE #2326: What namespaces to keep, what to drop, and what to hide
In the table below there are all the namespaces that are in ML.Net. 
They also all display in the docs site. Some of them, like Microsoft.ML.Ensemble needs to continue to exist in the code, but it is not ready to be exposed to the users, and can be hidden in the docs site. 

Some others, like Microsoft.ML.Internal.Internallearn can potentially be merged into other namespaces. 
Let's annotate in the list below the namespaces that need to be hidden in the docs site, and the ones that need to be gone altogether:

Namespace| Drop it from code | Hide it from docs | Ships In Nuget | 
-------- | ----------------- | ----------------- | ----------- |
Microsoft.ML |    |   | Microsoft.ML |
Microsoft.ML.Calibrator |    |   | Microsoft.ML  |
Microsoft.ML.Core.Data | drop   |   | 	  |
Microsoft.ML.Data |    |   | Microsoft.ML |	
Microsoft.ML.Data.IO | drop?   |   |	  |
Microsoft.ML.Data.IO.Zlib |  drop  |   |	  |
Microsoft.ML.Ensemble | Microsoft.ML.Trainers.Ensemble   |  hide |	None  |
Microsoft.ML.Ensemble.EntryPoints |    |   |Microsoft.ML  |	
Microsoft.ML.Ensemble.OutputCombiners |  Microsoft.ML.Trainers.Ensemble  | hide  |	 None |
Microsoft.ML.Ensemble.Selector | Microsoft.ML.Trainers.Ensemble   | hide  |	 None |
Microsoft.ML.Ensemble.Selector.DiversityMeasure | Microsoft.ML.Trainers.Ensemble   | hide  |	None  |
Microsoft.ML.EntryPoints |    |   |	Microsoft.ML  |
Microsoft.ML.EntryPoints.JsonUtils |  drop?  |   |  |	
Microsoft.ML.FactorizationMachine | Microsoft.ML.Trainers.FactorizationMachine   |   |	Microsoft.ML  |
Microsoft.ML.ImageAnalytics |    |   | Microsoft.ML.ImageAnalytics	  |
Microsoft.ML.ImageAnalytics.EntryPoints |   Microsoft.ML.ImageAnalytics |   |	  |
Microsoft.ML.Internal.Calibration | Microsoft.Ml.Calibrator   | hide  |	Microsoft.ML   |
Microsoft.ML.Internal.CpuMath |    |  hide |	Microsoft.ML   |
Microsoft.ML.Internal.Internallearn | drop?   |   |	  |
Microsoft.ML.Internal.Internallearn.ResultProcessor |   drop? |   | None	  |
Microsoft.ML.Learners |  Microsoft.ML.Trainers  |   | Microsoft.ML 	  |
Microsoft.ML.LightGBM | Microsoft.ML.Trainers.LightGBM   |   | Microsoft.ML.LightGBM	  |
Microsoft.ML.Model |    |   |	Microsoft.ML.Onnx ,  Microsoft.ML |
Microsoft.ML.Model.Onnx |    |   |	Microsoft.ML.Onnx  |
Microsoft.ML.Numeric |  drop?  |   |  |	
Microsoft.ML.SamplesUtils |    |   |	Microsoft.ML  |
Microsoft.ML.StaticPipe |    |   | Microsoft.ML.StaticPipe	  |
Microsoft.ML.StaticPipe.Runtime | drop?   |   |  |	
Microsoft.ML.Sweeper |    |  hide |	Microsoft.ML  |
Microsoft.ML.Sweeper.Algorithms |    |  hide | Microsoft.ML	  |
Microsoft.ML.TimeSeries |    |   |Microsoft.ML.TimeSeries	  |
Microsoft.ML.TimeSeriesProcessing |  drop?  |   |  |	
Microsoft.ML.Tools |    |   hide? |	  |
Microsoft.ML.Trainers |    |   |	Microsoft.ML  |
Microsoft.ML.Trainers.FastTree |    |   |	Microsoft.ML  |
Microsoft.ML.Trainers.FastTree.Internal |  Microsoft.ML.Trainers.FastTree|   |	  |
Microsoft.ML.Trainers.HalLearners |    |   Microsoft.ML.Trainers.HalLearners |	  |
Microsoft.ML.Trainers.KMeans |    |   |  Microsoft.ML |	
Microsoft.ML.Trainers.Online |    |   | Microsoft.ML  |	
Microsoft.ML.Trainers.PCA |    |   |	Microsoft.ML  |
Microsoft.ML.Trainers.Recommender |    |   |	  |
Microsoft.ML.Trainers.SymSgd |  Microsoft.ML.Trainers.HalLearners |   |  |	
Microsoft.ML.Training |  drop  |   |	  |
Microsoft.ML.Transforms |    |   |	  |
Microsoft.ML.Transforms.Categorical |    |   |	  |
Microsoft.ML.Transforms.Conversions |    |   |	  |
Microsoft.ML.Transforms.FeatureSelection |    |   |	  |
Microsoft.ML.Transforms.Normalizers |    |   |	  |
Microsoft.ML.Transforms.Projections |    |   |	  |
Microsoft.ML.Transforms.TensorFlow |    |   |	  |
Microsoft.ML.Transforms.Text |    |   |  |
Microsoft.ML.UniversalModelFormat.Onnx |  Microsoft.ML.Model.Onnx  |   |

### PR DESCRIPTION
**Pull request by sfilipi**
Closes #2326 by : 
1- First commit
substituting :

Microsoft.ML.Trainers.Ensemble.DiversityMeasure
Microsoft.ML.Trainers.Ensemble.FeatureSelector
Microsoft.ML.Trainers.Ensemble.SubModelSelector

with:
Microsoft.ML.Trainers.Ensemble

2 - Second commit:
Microsoft.ML.FactorizationMachine becomes Microsoft.ML.Trainers.FactorizationMachine
Microsoft.ML.EntryPoints.JsonUtils becomes Microsoft.ML.EntryPoints

3- third commit
Microsoft.ML.ImageAnalytics.EntryPoints becomes Microsoft.ML.ImageAnalytics
Microsoft.ML.Internal.Calibration becomes Microsoft.Ml.Calibrator

4- Fourth commit
Microsoft.ML.StaticPipe.Runtime changes to Microsoft.ML.StaticPipe

easier to review commit by commit. 

I have used the VS replace all + reordering namespaces through Ctrl+R+G

### CODE REVIEW

File: src/Microsoft.ML.FastTree/RandomForestClassification.cs [hunk 1]
```diff
@@ -10,7 +10,6 @@
 using Microsoft.ML.CommandLine;
 using Microsoft.ML.Data;
 using Microsoft.ML.EntryPoints;
-using Microsoft.ML.Internal.Calibration;
 using Microsoft.ML.Internal.Internallearn;
```
Comment (eerhardt) [inline src/Microsoft.ML.FastTree/RandomForestClassification.cs]: I assume `Internal.Internallearn` will be taken care of in another change.

Comment (codemzs) [reply src/Microsoft.ML.FastTree/RandomForestClassification.cs]: Yep, I’ll be taking care of that today.

File: src/Microsoft.ML.ImageAnalytics/EntryPoints/ImageAnalytics.cs [hunk 2]
```diff
@@ -4,10 +4,10 @@
...
 using Microsoft.ML;
 using Microsoft.ML.EntryPoints;
-using Microsoft.ML.ImageAnalytics.EntryPoints;
+using Microsoft.ML.ImageAnalytics;
 
 [assembly: LoadableClass(typeof(void), typeof(ImageAnalytics), null, typeof(SignatureEntryPointModule), "ImageAnalytics")]
-namespace Microsoft.ML.ImageAnalytics.EntryPoints
+namespace Microsoft.ML.ImageAnalytics
 {
     internal static class ImageAnalytics
```
Comment (wschin) [inline src/Microsoft.ML.ImageAnalytics/EntryPoints/ImageAnalytics.cs]: >ImageAnalytics [](start = 26, length = 14)

Humm, is it really good to have a class named by its namespace? Once I ran into missing-reference problem because of that.

File: src/Microsoft.ML.Data/Prediction/Calibrator.cs [hunk 3]
```diff
@@ -80,7 +79,7 @@
 [assembly: EntryPointModule(typeof(PavCalibratorTrainerFactory))]
 [assembly: EntryPointModule(typeof(PlattCalibratorTrainerFactory))]
 
-namespace Microsoft.ML.Internal.Calibration
+namespace Microsoft.ML.Calibrator
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Prediction/Calibrator.cs]: >Calibrator [](start = 23, length = 10)

If we really want to have another namespace that's fine I guess, but if the namespace name is `Microsoft.ML.Trainers` then this should probably be `Microsoft.ML.Calibrators`, not `Calibrator`.

**Comment by TomFinley** [comment]
> namespace Microsoft.ML.Calibrator

Similar here.

---
Refers to: src/Microsoft.ML.Data/Prediction/CalibratorCatalog.cs:23 in a0eb13d. [](commit_id = a0eb13dc4508f9349e77ff82d02a99addf9c0d29, deletion_comment = False)

**Comment by singlis** [comment]
> namespace Microsoft.ML.LightGBM.StaticPipe

Should the namespace for this be Microsoft.ML.StaticPipe? Looking at other examples, classes are using Microsoft.ML.StaticPipeline for the namespace.

---
Refers to: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs:11 in 2af661d. [](commit_id = 2af661d3569c341000127301e56b9f20cb258e94, deletion_comment = False)

**Comment by singlis** [comment]
> namespace Microsoft.ML.LightGBM.StaticPipe

Are there other classes in the StaticPipe that don't use Microsoft.ML.StaticPipe?


---
In reply to: [467609030](https://github.com/dotnet/machinelearning/pull/2724#issuecomment-467609030) [](ancestors = 467609030)

---
Refers to: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs:11 in 2af661d. [](commit_id = 2af661d3569c341000127301e56b9f20cb258e94, deletion_comment = False)

**Comment by singlis** [comment]
> namespace Microsoft.ML.LightGBM.StaticPipe

Some other variations that I found:
Microsoft.ML.StaticPipe.Runtime
Microsoft.ML.HalLearners.StaticPipe
Microsoft.ML.Transforms.StaticPipe


---
In reply to: [467609182](https://github.com/dotnet/machinelearning/pull/2724#issuecomment-467609182) [](ancestors = 467609182,467609030)

---
Refers to: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs:11 in 2af661d. [](commit_id = 2af661d3569c341000127301e56b9f20cb258e94, deletion_comment = False)

**Comment by singlis** [comment]
Looks good to me, I left some feedback.

File: src/Microsoft.ML.Data/Prediction/Calibrator.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Data/Prediction/Calibrator.cs]: Should we do `Microsoft.ML.StaticPipe"s"`?

**Comment by sfilipi** [comment]
> namespace Microsoft.ML.LightGBM.StaticPipe

let me address that on a separate issue/PR since it is not related to the core Microsoft.ML nuget. 

Logged: [https://github.com/dotnet/machinelearning/issues/2746](https://github.com/dotnet/machinelearning/issues/2746)

---
In reply to: [467610688](https://github.com/dotnet/machinelearning/pull/2724#issuecomment-467610688) [](ancestors = 467610688,467609182,467609030)

---
Refers to: src/Microsoft.ML.LightGBM.StaticPipe/LightGbmStaticExtensions.cs:11 in 2af661d. [](commit_id = 2af661d3569c341000127301e56b9f20cb258e94, deletion_comment = False)

File: src/Microsoft.ML.Data/Prediction/Calibrator.cs (hunk 3 above)
Comment (sfilipi) [reply src/Microsoft.ML.Data/Prediction/Calibrator.cs]: let's do that as part of: 
[https://github.com/dotnet/machinelearning/issues/2746](https://github.com/dotnet/machinelearning/issues/2746)

---
In reply to: [260485716](https://github.com/dotnet/machinelearning/pull/2724#discussion_r260485716) [](ancestors = 260485716)

**Comment by eerhardt** [comment]
@Ivanidzo4ka @sfilipi - 

From our OneNote yesterday we have:

```
	5- Microsoft.ML.FastTree -> Going to Microsoft.ML.Trainers.FastTree :QuantileRegressionTree, QuantileRegressionTreeEnsemble, RegressionTree, RegressionTreeBase, RegressionTreeEnsemble, TreeEnsemble -->
done PR: https://github.com/dotnet/machinelearning/pull/2724
```

I don't see that happening in this PR. Should we open a new issue for that task, since it wasn't done here?
